README.md

[![Arduino CI](https://github.com/RobTillaart/SHEX/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/SHEX/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/SHEX/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/SHEX/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/SHEX/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/SHEX/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/SHEX.svg?maxAge=3600)](https://github.com/RobTillaart/SHEX/releases)


# SHEX

Arduino library to generate hex dump over Serial (any stream).


## Description

### SHEX

**SHEX** is a simple library that wraps the Serial output side (by default) and
generates an hex dump of all data that is printed. 16 bytes per row.

The default output format is (changed since 0.3.0)
```
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx

ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx 
```

with a separator line after each 8th line.

The constructor has a length parameter which can be used to have another number of bytes per row.
This can be changed with **setBytesPerLine()**.

One can toggle HEX output or pass through by means of **setHEX(bool)**.
This makes it possible to switch between the modes e.g. between 'debugging' and 'release' mode.

One can toggle the character count at the start of the line.


### SHEXA

**SHEXA** (Serial HEX ASCII) is a derived class from **SHEX** that also 
displays a column with printable characters.


The default output format is 
```
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx  abcdefgh ijklmnop
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx  abcdefgh ijklmnop
ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx  abcdefgh ijklmnop

ABCD  xx xx xx xx  xx xx xx xx  xx xx xx xx  xx xx xx xx  etc
```

To print this ASCII column extra RAM and code is used. 
Therefore this is made a derived class from **SHEX**.

Furthermore **SHEXA** has a function **flushASCII()** to flush the ASCII column to output.
This is needed when HEX output is restarted.


## Interface

```cpp
#include "SHEX.h"
```

The **SHEX** and **SHEXA** share most of their interface.

### Defines

To be adjusted via command line (or in SHEX.h file)

|  define               |  value  |
|:----------------------|--------:|
|  SHEX_DEFAULT_LENGTH  |     16  |
|  SHEX_MAX_LENGTH      |     32  |
|  SHEX_MIN_LENGTH      |      4  |
|  SHEX_COUNTER_DIGITS  |      4  |
|  SHEX_DEFAULT_VTAB    |      8  |


### Constructor + Core

- **SHEX(Print \* stream = &Serial, uint8_t length = SHEX_DEFAULT_LENGTH)** Constructor, 
optional set the number of bytes per line.
default 16 bytes per line, forced multiple of 4, max SHEX_MAX_LENGTH = 32.
- **size_t write(uint8_t c)** implements the Print interface.


### Output Modifiers

- **void setHEX(bool hexOutput = true)** switch between modi, HEX (true) or pass through (false).
- **bool getHEX()** returns mode set above.
- **void setBytesPerLine(uint8_t length = SHEX_DEFAULT_LENGTH)** idem, default 16 bytes per line, 
forced multiple of 4, max SHEX_MAX_LENGTH = 32.
- **uint8_t getBytesPerLine()** returns number of bytes per line.


### Separator

- **void setSeparator(char c = ' ')** set the separator character, default a space.
Some people like a dot '.', or a tab '\t'. Feel free to experiment.
- **char getSeparator()** return the separator character set.


### Counter

- **void setCountDigits(uint8_t digits)** set the length of the counter, 8 or 6 or 4 (default). 
Other numbers will be rounded up to 4, 6 or 8.
- **uint8_t getCountDigits()** returns idem.
- **void restartOutput()** restarts the counter from 0 and a new line.
Is automatically called if a setting is modified like **bytesPerLine**
**setVTAB** or **setCountDigits**
- **uint32_t getCounter()** return internal counter.


### VTAB

- **void setVTAB(uint8_t vtab = SHEX_DEFAULT_VTAB)** set the vertical separator line. 
- **uint8_t getVTAB()** return the current vertical separator line.


### SHEXA specific

- **void flushASCII()** allows the user to flush the ASCII column to output.
This is typically used after a setting is changed, which causes a restart of
the HEX output. Best if followed by a **restartOutput()**.
_Note: it is not ideal but workable. This might change in the future._


## Operational

See examples.


## Future

Although no follow up release is planned, some ideas are kept here
so they won't get lost.


#### Must

- optimize code
  - print vs write


#### Should

- more testing
  - performance measurement
  - different platforms. 
  - different streams incl SW Serial


#### Could

- investigate **flushASCII()** performance.
- HEX reader: **RHEX** converts dump format to a normal stream again.
  - separate library.
- void setSeparatorInterval()  ??


#### Wont

- line buffering for faster output (e.g Ethernet and SD card)
  could it support **write(array, length)** call.
  needs quite a rewrite..
  Needs a big buffer: ~150 bytes. (counter 8 + hex 96 + printable 32 + extra sep)
  ==> go through a small buffer twice as data is same!!
- header line: runtime configurable; 
  optional combined with separator
  and after how many lines the header should repeat
  **header(str, lines)** ?
  Can also be done by just calling Serial.print.


CHANGELOG.md
# Change Log SHEX

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.2] - 2023-02-20
- fix **floshASCII()**
- fix LIB version number.
- fix command line defines.
- move code from .h to .cpp
- update readme.md
- update GitHub actions
- update license 2023
- minor edits


## [0.3.1] - 2022-11-24
- Add RP2040 support to build-CI.
- Add CHANGELOG.md

## [0.3.0   2022-05-28  breaking!
- change default HEX output instead of pass through.
- add get / setCountDigits() =>
-     #digits of count 4, 6 or 8 (4 = default)
- replaces get / setCounterFlag()
- add define SHEX_COUNTER_DIGITS + SHEX_MIN_LENGTH
- add restartOutput() and getCounter()
- add SHEXA class for ASCII column output
- add SHEXA::flushASCII().

----

## [0.2.3] - 2022-05-28
- add setVTAB(vtab) getVTAB()
- add define SHEX_DEFAULT_VTAB

## [0.2.2] - 2022-05-27
- fix #6 set default length
- add defines SHEX_DEFAULT_LENGTH + SHEX_MAX_LENGTH

## [0.2.1] - 2021-12-28
- update library.json
- update readme.md
- update license
- minor edits

## [0.2.0] - 2021-01-07
- Arduino-CI + unit tests + modifiers

----

## [0.1.1] - 2020-06-19
- fix library.json

## [0.1.0] - 2020-05-24
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-01-07
// PURPOSE: unit tests for the SHEX Serial HEX library
//          https://github.com/RobTillaart/SHEX
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>

#include "SHEX.h"


unittest_setup()
{
  fprintf(stderr, "SHEX_LIB_VERSION: %s\n", (char *) SHEX_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(16, SHEX_DEFAULT_LENGTH);
  assertEqual(32, SHEX_MAX_LENGTH);
  assertEqual(4, SHEX_MIN_LENGTH);
  assertEqual(4, SHEX_COUNTER_DIGITS);
  assertEqual(8, SHEX_DEFAULT_VTAB);
}


unittest(test_set_hex)
{
  SHEX shex;

  assertTrue(shex.getHEX());

  shex.setHEX(false);
  assertFalse(shex.getHEX());

  shex.setHEX();
  assertTrue(shex.getHEX());
}


unittest(test_bytes_per_line)
{
  SHEX shex(&Serial, 24);

  assertEqual(24, shex.getBytesPerLine());

  shex.setBytesPerLine(8);
  assertEqual(8, shex.getBytesPerLine());

  shex.setBytesPerLine(60);
  assertEqual(32, shex.getBytesPerLine());

  shex.setBytesPerLine();
  assertEqual(SHEX_DEFAULT_LENGTH, shex.getBytesPerLine());
}


unittest(test_separator)
{
  SHEX shex;

  assertEqual(' ', shex.getSeparator());

  shex.setSeparator('-');
  assertEqual('-', shex.getSeparator());

  shex.setSeparator();
  assertEqual(' ', shex.getSeparator());
}


unittest(test_count_digits)
{
  SHEX shex;

  assertEqual(SHEX_COUNTER_DIGITS, shex.getCountDigits());

  for (uint8_t digit = 0; digit < 9; digit++)
  {
    shex.setCountDigits(digit);
    assertEqual(digit, shex.getCountDigits());
  }

  //  should be truncated to 8 == uint32_t counter)
  shex.setCountDigits(9);
  assertEqual(8, shex.getCountDigits());
}


unittest(test_vtab)
{
  SHEX shex;

  assertEqual(8, shex.getVTAB());

  for (uint8_t vtab = 1; vtab < 9; vtab *= 2 )
  {
    shex.setVTAB(vtab);
    assertEqual(vtab, shex.getVTAB());
  }
}


unittest_main()

// --------


SHEX.cpp
//
//    FILE: SHEX.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.3.2
// PURPOSE: Arduino library to generate hex dump over Serial
//    DATE: 2020-05-24
//     URL: https://github.com/RobTillaart/SHEX


#include "SHEX.h"


SHEX::SHEX(Print* stream, uint8_t length)
{
  _stream  = stream;
  reset();
  //  force multiple of 4;  max 32
  _length = ((length + 3) / 4) * 4;
  if (_length > SHEX_MAX_LENGTH)
  {
    _length = SHEX_MAX_LENGTH;
  }
  if (_length < SHEX_MIN_LENGTH)
  {
    _length = SHEX_MIN_LENGTH;
  }
}


void SHEX::reset()
{
  _hexOutput = true;
  _length    = SHEX_DEFAULT_LENGTH;
  _charCount = 0;
  _separator = ' ';
  _digits    = SHEX_COUNTER_DIGITS;
  _vtab      = SHEX_DEFAULT_VTAB;
}


///////////////////////////////////////////
//
//  WRITE - the core
//
size_t SHEX::write(uint8_t c)
{
  //  PASS THROUGH MODE
  if (_hexOutput == false) return _stream->write(c);

  //  HEX MODE
  //  handle end of line and position number
  if ((_charCount % _length) == 0)
  {
    //  insert ASCII array here

    _stream->println();
    //  separator line every _vtab (default 8) lines
    if ((_charCount % (_length * _vtab)) == 0)
    {
      _stream->println();
    }

    //  next line
    if (_digits > 0)
    {
      uint32_t mask = 0xF000;
      if (_digits > 4) mask = 0xF00000;
      if (_digits > 6) mask = 0xF0000000;
      while((mask > 0xF) && (mask & _charCount) == 0)
      {
        _stream->print('0');
        mask >>= 4;
      }
      _stream->print(_charCount, HEX);
      _stream->print('\t');
    }
  }

  //  Print char as HEX
  if (c < 0x10) _stream->print('0');
  _stream->print(c, HEX);
  _stream->print(_separator);
  _charCount++;
  if ((_charCount % 4) == 0) _stream->print(_separator);

  return 1;
}


void SHEX::setHEX(bool hexOutput)
{
  _hexOutput = hexOutput;
  restartOutput();
}


bool SHEX::getHEX()
{
  return _hexOutput;
}


void SHEX::setBytesPerLine(const uint8_t length)
{
  //  force multiple of 4; max 32
  _length = ((length + 3) / 4) * 4;
  if (_length > SHEX_MAX_LENGTH)
  {
    _length = SHEX_MAX_LENGTH;
  }
  if (_length < SHEX_MIN_LENGTH)
  {
    _length = SHEX_MIN_LENGTH;
  }
  restartOutput();
}


uint8_t SHEX::getBytesPerLine()
{
  return _length;
}


void SHEX::setSeparator(char c = ' ')
{
  _separator = c;
}


char SHEX::getSeparator()
{
  return _separator;
}


uint8_t SHEX::getCountDigits()
{
  return _digits;
}


uint32_t SHEX::getCounter()
{
  return _charCount;
}


void SHEX::setVTAB(uint8_t vtab)
{
  _vtab = vtab;
  restartOutput();
};


uint8_t SHEX::getVTAB()
{
  return _vtab;
}


void SHEX::setCountDigits(uint8_t digits)
{
  _digits = digits;
  if (_digits == 0) return;
  if (_digits < 4) _digits = 4;
  if (_digits > 8) _digits = 8;
  restartOutput();
};


void SHEX::restartOutput()
{
  //  prevent change in middle of line
  _charCount = 0;
  _stream->println();
}


///////////////////////////////////////////////////
//
//  SHEXA
//
SHEXA::SHEXA(Print* stream, uint8_t length) : SHEX(stream, length)
{
}


size_t SHEXA::write(uint8_t c)
{
  //  PASS THROUGH MODE
  if (_hexOutput == false) return _stream->write(c);

  //  HEX MODE
  //  handle end of line and position number
  if ((_charCount % _length) == 0)
  {
    //  printable ASCII column
    if (_charCount != 0) 
    {
      flushASCII();
    }

    _stream->println();
    //  separator line every _vtab (default 8) lines
    if ((_charCount % (_length * _vtab)) == 0)
    {
      _stream->println();
    }

    //  next line
    if (_digits > 0)
    {
      uint32_t mask = 0xF000;
      if (_digits > 4) mask = 0xF00000;
      if (_digits > 6) mask = 0xF0000000;
      while((mask > 0xF) && (mask & _charCount) == 0)
      {
        _stream->print('0');
        mask >>= 4;
      }
      _stream->print(_charCount, HEX);
      _stream->print('\t');
    }
  }

  //  Print char as HEX
  if (c < 0x10) _stream->print('0');
  _stream->print(c, HEX);
  _stream->print(_separator);

  //  Store in _txtbuf
  _txtbuf[_charCount % _length] = isPrintable(c) ? c : '.';

  _charCount++;
  if ((_charCount % 4) == 0) _stream->print(_separator);

  return 1;
}


void SHEXA::flushASCII()
{
  uint8_t len = _charCount % _length;
  if (len == 0)
  {
    len = _length;
  }
  if (len < _length)
  {
    //  merge in one loop
    for (uint8_t i = len; i < _length; i++)
    {
      _stream->print("   ");
    }
    for (uint8_t i = 0; i < (_length/4 - len/4); i++)
    {
      _stream->print(" ");
    }
  }

  for (uint8_t i = 0; i < len;)
  {
    _stream->write(_txtbuf[i++]);
    if ((i % 8) == 0)
    {
      _stream->print("  ");
    }
  }
}


//  -- END OF FILE --



SHEX.h
#pragma once
//
//    FILE: SHEX.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.3.0
// PURPOSE: Arduino library to generate hex dump over Serial
//    DATE: 2020-05-24
//     URL: https://github.com/RobTillaart/SHEX


#include "Arduino.h"
#include "Print.h"


#define SHEX_LIB_VERSION                (F("0.3.0"))

#define SHEX_DEFAULT_LENGTH             16
#define SHEX_MAX_LENGTH                 32
#define SHEX_MIN_LENGTH                 4
#define SHEX_COUNTER_DIGITS             4
#define SHEX_DEFAULT_VTAB               8


class SHEX: public Print
{
public:
  SHEX(Print* stream = &Serial, uint8_t length = SHEX_DEFAULT_LENGTH);

  void    reset();

  size_t  write(uint8_t c);

  void    setHEX(bool hexOutput = true);
  bool    getHEX() { return _hexOutput; };

  void    setBytesPerLine(const uint8_t length = SHEX_DEFAULT_LENGTH);
  uint8_t getBytesPerLine() { return _length; };

  void    setSeparator(char c = ' ') { _separator = c; };
  char    getSeparator() { return _separator; };

  //      must be 0, 4, 6 or 8
  void    setCountDigits(uint8_t digits = SHEX_COUNTER_DIGITS);
  uint8_t getCountDigits() { return _digits; }

  void    setVTAB(uint8_t vtab = SHEX_DEFAULT_VTAB);
  uint8_t getVTAB() { return _vtab; };


protected:
  Print *   _stream    = &Serial;
  bool      _hexOutput = true;
  uint8_t   _length    = SHEX_DEFAULT_LENGTH;
  uint8_t   _vtab      = SHEX_DEFAULT_VTAB;
  char      _separator = ' ';

  uint32_t  _charCount = 0;
  uint32_t  _digits    = SHEX_COUNTER_DIGITS;
};


///////////////////////////////////////////////////
//
//  SHEXA shows also ASCII dump
//        derived class as it takes extra RAM
//
class SHEXA : public SHEX
{
public:
  SHEXA(Print* stream = &Serial, uint8_t length = SHEX_DEFAULT_LENGTH);

  size_t    write(uint8_t c);
  //  flushes the ASCII column to output;  not ideal but workable for now.
  //  use with care
  void      flushASCII();

protected:
  uint8_t   _txtbuf[SHEX_MAX_LENGTH];
  bool      _showtxt = true;
};


// -- END OF FILE --

